sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -1054,11 +1056,18 @@ private[hive] object HiveClientImpl {
     // When table is external, `totalSize` is always zero, which will influence join strategy.
     // So when `totalSize` is zero, use `rawDataSize` instead. When `rawDataSize` is also zero,
     // return None.
+    // If a table has a deserialization factor, the table owner expects the in-memory
+    // representation of the table to be larger than the table's totalSize value. In that case,
+    // multiply totalSize by the deserialization factor and use that number instead.
+    // If the user has set spark.sql.statistics.ignoreRawDataSize to true (because of HIVE-20079,
+    // for example), don't use rawDataSize.
     // In Hive, when statistics gathering is disabled, `rawDataSize` and `numRows` is always
     // zero after INSERT command. So they are used here only if they are larger than zero.
-    if (totalSize.isDefined && totalSize.get > 0L) {
-      Some(CatalogStatistics(sizeInBytes = totalSize.get, rowCount = rowCount.filter(_ > 0)))
-    } else if (rawDataSize.isDefined && rawDataSize.get > 0) {
+    val adjustedSize = DataSourceUtils.calcDataSize(properties, totalSize.getOrElse(BigInt(0)))
+    val sqlConf = SQLConf.get
+    if (adjustedSize > 0L) {
+      Some(CatalogStatistics(sizeInBytes = adjustedSize, rowCount = rowCount.filter(_ > 0)))
+    } else if (rawDataSize.isDefined && rawDataSize.get > 0 && !sqlConf.ignoreRawDataSize) {
```

Comment:
@bersprockets  Here can we calculate the raw size based on the row count instead of use the not use the raw size? Thanks.